CAP/gap/DerivedMethods.gi

```diff
@@ -3329,3 +3329,159 @@ AddFinalDerivation( IsEqualForCacheForMorphisms,
                     
   IsIdenticalObj : Description := "Use IsIdenticalObj for IsEqualForCacheForMorphisms" );
+
+## Final methods for BasisOfExternalHom & CoefficientsOfMorphism
+
+##
+AddFinalDerivation( BasisOfExternalHom,
+                    [
+                      [ HomomorphismStructureOnObjects ],
+                      [ HomomorphismStructureOnMorphismsWithGivenObjects ],
+                      [ InterpretMorphismFromDistinguishedObjectToHomomorphismStructureAsMorphism ],
+                      [ InterpretMorphismAsMorphismFromDistinguishedObjectToHomomorphismStructure ],
+                      [ DistinguishedObjectOfHomomorphismStructure ],
+                      [ MultiplyWithElementOfCommutativeRingForMorphisms ]
+                    ],
+                    [
+                      BasisOfExternalHom,
+                      CoefficientsOfMorphismWithGivenBasisOfExternalHom
+                    ],
+  function( a, b )
+    local hom_a_b, D, B, ConditionsListComplete;
+    
+    hom_a_b := HomomorphismStructureOnObjects( a, b );
+    
+    D := DistinguishedObjectOfHomomorphismStructure( CapCategory( a ) );
+    
+    B := ValueGlobal( "BasisOfExternalHom" )( D, hom_a_b );
+    
+    return List( B, m -> InterpretMorphismFromDistinguishedObjectToHomomorphismStructureAsMorphism( a, b, m ) );
+  
+end : ConditionsListComplete := true,
+  CategoryFilter :=
+    function( category )
+      local range_cat, required_methods;
+      
+      if not ( HasIsLinearCategoryOverCommutativeRing( category ) and IsLinearCategoryOverCommutativeRing( category ) ) then
+        
+        return false;
+        
+      fi;
+      
+      if not HasRangeCategoryOfHomomorphismStructure( category ) then
+        
+        return false;
+        
+      fi;
+      
+      range_cat := RangeCategoryOfHomomorphismStructure( category );
+      
+      if IsIdenticalObj( category, range_cat ) then
+        
+        return false;
+        
+      fi;
+      
+      if not ( HasIsLinearCategoryOverCommutativeRing( range_cat ) and IsLinearCategoryOverCommutativeRing( range_cat ) ) then
+        
+        return false;
+        
+      fi;
+      
+      if not IsIdenticalObj( CommutativeRingOfLinearCategory( category ), CommutativeRingOfLinearCategory( range_cat ) ) then
+        
+        return false;
+        
+      fi;
+      
+      required_methods := [
+                            "BasisOfExternalHom",
+                            "CoefficientsOfMorphismWithGivenBasisOfExternalHom",
+                            "MultiplyWithElementOfCommutativeRingForMorphisms"
+                          ];
+      
+      if not ForAll( required_methods, m -> CanCompute( range_cat, m ) ) then
+        
+        return false;
+        
+      fi;
+      
+      return true;
+      
+  end,
+  Description := "Adding BasisOfExternalHom using homomorphism structure" );
+
+##
+AddFinalDerivation( CoefficientsOfMorphismWithGivenBasisOfExternalHom,
```

Comment:
I think it's better to bundle this final derivation and the one for `BasisOfExternalHom` as a "final derivation package", like it was done for the weak kernel here:
https://github.com/homalg-project/CAP_project/blob/7a2391216c737a2e9cc004d0dc1ca8f7bf541682/FreydCategoriesForCAP/gap/FreydCategoriesDerivedMethods.gi#L282

If you don't bundle it, I fear that it is not possible that both final derivations can be triggered at the same time.

Reply:
Very nice. I have updated the derivation.